api/agent/agent.go

```diff
@@ -334,52 +341,52 @@ func (a *agent) hotLauncher(ctx context.Context, callObj *call) {
 
 		curStats := callObj.slots.getStats()
 		isNeeded := isNewContainerNeeded(&curStats)
-		logger.WithFields(logrus.Fields{
-			"currentStats": curStats,
-			"isNeeded":     isNeeded,
-		}).Debug("Hot function launcher stats")
+		logger.WithFields(logrus.Fields{"currentStats": curStats, "isNeeded": isNeeded}).Debug("Hot function launcher stats")
 		if !isNeeded {
 			continue
 		}
+		logger.WithFields(logrus.Fields{"currentStats": curStats, "isNeeded": isNeeded}).Info("Hot function launcher starting hot container")
 
-		ctxResource, cancelResource := context.WithCancel(context.Background())
-		logger.WithFields(logrus.Fields{
-			"currentStats": curStats,
-			"isNeeded":     isNeeded,
-		}).Info("Hot function launcher starting hot container")
+		// if we get a timeout, we must call off get resource token
+		ctx, cancel := context.WithCancel(ctx)
 
 		select {
-		case tok, isOpen := <-a.resources.GetResourceToken(ctxResource, callObj.Memory, uint64(callObj.CPUs), isAsync):
-			cancelResource()
+		case tok, isOpen := <-a.resources.GetResourceToken(ctx, callObj.Memory, uint64(callObj.CPUs), isAsync):
+			cancel()
 			if isOpen {
 				a.wg.Add(1)
 				go func(ctx context.Context, call *call, tok ResourceToken) {
 					a.runHot(ctx, call, tok)
 					a.wg.Done()
 				}(ctx, callObj, tok)
 			} else {
-				// this means the resource was impossible to reserve (eg. memory size we can never satisfy)
+				// this means the resource was impossible to reserve (eg. memory size we can never satisfy),
+				// return here so we don't infinite loop
 				callObj.slots.queueSlot(&hotSlot{done: make(chan struct{}), err: models.ErrCallTimeoutServerBusy})
+				return
```

Comment:
we can punt on this for now and fix this later as well, but we can't return here without deleting the slot. I think for now, we can just queue the error without a return or maybe we could move this as a validation check much before we even bother launching this hot slot. We could probably call isResourcePossible() directly before even attempting getSlotQueue in getSlot() and return an err immediately..

Reply:
i've made some re-arrangements of the fine china once again https://github.com/fnproject/fn/pull/716/commits/c4374de0dbab92fe8a3e3d78587de2cf883c92c7 -- hopefully, there are no cracked teapots this time. GetResourceToken contract was kind of bugging me, now there is no possibility of us queueing an error slot in here since GetCall will block anybody that will try to ask for too much